src/shared/codelens/localLambdaRunner.ts

```diff
@@ -532,72 +531,67 @@ const getEnvironmentVariables = (config: HandlerConfig): SAMTemplateEnvironmentV
+    onStartDebugging?: typeof vscode.debug.startDebugging
+    onRecordAttachDebuggerMetric?(attachResult: boolean | undefined, attempts: number, attachResultDate: Date): void
+    onWillRetry?(): Promise<void>
+}
+
+export async function attachDebugger(
+    {
+        retryDelayMillis = ATTACH_DEBUGGER_RETRY_DELAY_MILLIS,
+        onStartDebugging = vscode.debug.startDebugging,
+        onWillRetry = async (): Promise<void> => {
+            await new Promise<void>(resolve => {
+                setTimeout(resolve, retryDelayMillis)
+            })
+        },
+        ...params
+    }: AttachDebuggerContext
+): Promise<{ success: boolean }> {
+    const channelLogger = params.channelLogger
+    const logger = params.channelLogger.logger
     logger.debug(`localLambdaRunner.attachDebugger: startDebugging with debugConfig: ${JSON.stringify(
         params.debugConfig,
         undefined,
         2
     )}`)
 
-    let isDebuggerAttached: boolean | undefined = false
-    let numAttempts = 0
-    let retryDelay = 1000
-    let shouldRetry = false
-    const retryEnabled = false // Change this to enable retry
+    let isDebuggerAttached: boolean | undefined
+    let retries = 0
+
+    channelLogger.info(
+        'AWS.output.sam.local.attaching',
+        'Attaching debugger to SAM Application...',
+    )
+
     do {
-        channelLogger.info(
-            'AWS.output.sam.local.attaching',
-            'Attempt number {0} to attach debugger to SAM Application...',
-            String(numAttempts + 1)
-        )
-        isDebuggerAttached = await vscode.debug.startDebugging(undefined, params.debugConfig)
-        numAttempts += 1
-        if (isDebuggerAttached === undefined) {
-            isDebuggerAttached = false
-            shouldRetry = numAttempts < MAX_DEBUGGER_ATTEMPTS
-        } else if (!isDebuggerAttached) {
-            retryDelay *= 2
-
-            shouldRetry = retryEnabled && (numAttempts < MAX_DEBUGGER_ATTEMPTS)
-            if (shouldRetry) {
-                const currTime = new Date()
-                recordDebugAttachResult({
-                    telemetryService: params.telemetryService,
-                    attachResult: isDebuggerAttached,
-                    attempts: numAttempts,
-                    duration: currTime.getTime() - params.startInvokeMillis,
-                    runtime: params.runtime,
-                })
+        isDebuggerAttached = await onStartDebugging(undefined, params.debugConfig)
 
-                channelLogger.info(
-                    'AWS.output.sam.local.attach.retry',
-                    'Will try to attach debugger again in {0} seconds...',
-                    String(retryDelay / 1000)
+        if (isDebuggerAttached === undefined) {
+            if (retries < params.maxRetries) {
+                if (onWillRetry) {
+                    await onWillRetry()
+                }
+                retries += 1
+            } else {
+                channelLogger.error(
```

Comment:
Alternative to consider: instead of allowing maxRetries to be configurable, put up a message prompt that reads along the lines of "VS Code is having trouble attaching the debugger" with a retry and cancel button, with retry adding 10 to the params.maxRetries amount and cancelling going down the current `else` path.

Reply:
I like your suggestion as a possible addition to the configurable retry. How about we put it into the backlog?

Reply:
Done: https://github.com/aws/aws-toolkit-vscode/issues/446